src/include/ikarus/utils/PathFollowingFunctions.hh

```diff
@@ -0,0 +1,143 @@
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
+ */
+
+#pragma once
+#include <memory>
+
+#include <dune/common/float_cmp.hh>
+
+#include "ikarus/solver/linearSolver/linearSolver.hh"
+
+namespace Ikarus {
+
+  struct SubsidiaryArgs {
+    double stepSize;
+    Eigen::VectorX<double> DD;
+    double Dlambda{};
+    double f{};
+    Eigen::VectorX<double> dfdDD;
+    double dfdDlambda{};
+  };
+
+  /// Arc Length Control Method
+  struct StandardArcLength {
+  public:
+    explicit StandardArcLength(double p_psi = 1.0) : psi{p_psi} {}
+
+    void evaluateSubsidiaryFunction(SubsidiaryArgs& args) const {
+      args.f          = sqrt(args.DD.squaredNorm() + psi * psi * args.Dlambda * args.Dlambda) - args.stepSize;
+      args.dfdDD      = args.DD / (args.f + args.stepSize);
+      args.dfdDlambda = (psi * psi * args.Dlambda) / (args.f + args.stepSize);
+
+      if (Dune::FloatCmp::eq((args.f + args.stepSize), 0.0)) {
+        // During initialisation of the solver, (f + stepSize = 0)
+
+        args.dfdDlambda = 0.0;
+        args.dfdDD.setZero();
+      }
+    }
+
+    template <typename NonLinearOperator>
+    void initialPrediction(NonLinearOperator& nonLinearOperator, SubsidiaryArgs& args) {
+      auto linearSolver
+          = Ikarus::ILinearSolver<double>(Ikarus::SolverTypeTag::d_LDLT);  // for the linear predictor step
+
+      nonLinearOperator.lastParameter() = 1.0;  // lambda =1.0
+      psi                               = 1.0;
+      nonLinearOperator.template update<0>();
+      const auto& R = nonLinearOperator.value();
+      const auto& K = nonLinearOperator.derivative();
+
+      linearSolver.factorize(K);
+      linearSolver.solve(args.DD, -R);
+
+      const auto DD2 = args.DD.squaredNorm();
+
+      psi    = sqrt(DD2);
+      auto s = sqrt(psi * psi + DD2);
+
+      args.DD      = args.DD * args.stepSize / s;
+      args.Dlambda = args.stepSize / s;
+
+      nonLinearOperator.firstParameter() = args.DD;
+      nonLinearOperator.lastParameter()  = args.Dlambda;
+    }
+
+    template <typename NonLinearOperator>
+    void intermediatePrediction(NonLinearOperator& nonLinearOperator, SubsidiaryArgs& args) {
+      nonLinearOperator.firstParameter() += args.DD;
+      nonLinearOperator.lastParameter() += args.Dlambda;
+    }
+
+  private:
+    double psi;
+  };
+
+  /// Load Control Method
+  struct LoadControlWithSubsidiaryFunction {
+  public:
+    void evaluateSubsidiaryFunction(SubsidiaryArgs& args) const {
+      args.f = args.Dlambda - args.stepSize;
+      args.dfdDD.setZero();
```

Comment:
It is not really nice, that we assume that we receive the correctly resized `dfdDD`